feat: add support for JSON data type

olavloite opened the pull request:
Adds support for `JSON` data type.

olavloite committed:
feat: support JSON data type

olavloite committed:
fix: remove unnecessary changes

olavloite commented on samples/json-update-data.js:
```diff
@@ -0,0 +1,137 @@
+      // array of objects that should be inserted into a JSON column. If we were
+      // to specify this value as an array instead of a string, the client
+      // library would encode this value as ARRAY<JSON> instead of JSON.
+      VenueDetails: `[
```
@zoercai Here's an example of the issue with top-level JSON arrays. The client library does not know whether this should be encoded as `JSON` or `ARRAY<JSON>` if it is specified as an actual array instead of a string. Specifying it as a string will cause the client library to just encode it as a string, which is the default encoding for `JSON` values.

olavloite commented on src/codec.ts:
```diff
@@ -418,6 +422,10 @@ function encodeValue(value: Value): Value {
     return value.map(encodeValue);
   }
 
+  if (is.object(value)) {
+    return JSON.stringify(value);
+  }
```
This is where the most important part of this PR is: It ensures that any value that has been specified as an object will be converted to a valid `JSON` string. An array of objects will be converted to an array of valid `JSON` strings.

olavloite committed:
fix: linting + skip tests on emulator

olavloite committed:
docs: simplify samples

olavloite committed:
feat: decorate commit error with additional information

Decorate the error returned by a commit RPC in the specific case that
the application tries to insert an array of objects into a JSON column.

olavloite commented on src/transaction.ts:
```diff
@@ -1560,12 +1561,97 @@ export class Transaction extends Dml {
             resp.commitTimestamp as DateStruct
           );
         }
+        err = Transaction.decorateCommitError(err as ServiceError, mutations);
```
@skuruppu This adds additional information to the error for the specific case that the application tries to insert an array of objects into a JSON column.

skuruppu replied:
Perfect, thanks for doing this. If you could also please add some unit tests for this, that would be awesome.

olavloite replied:
Good point, added.

skuruppu replied:
Thanks a lot for adding the tests.

olavloite committed:
test: add unit tests for decorateError

skuruppu commented:
@olavloite I just merged in the codegen PR so you can rebase this PR now.

olavloite commented:
I added the Do not merge label back as the system test is failing with

```
  1) Spanner
       types
         "before all" hook for "should throw an error for incorrect value types":
     Error: 12 UNIMPLEMENTED: JSON type is not enabled.
      at Object.callErrorFromStatus (node_modules/@grpc/grpc-js/build/src/call.js:31:26)
          -> /workspace/node_modules/@grpc/grpc-js/src/call.ts:81:24
      at Object.onReceiveStatus (node_modules/@grpc/grpc-js/build/src/client.js:179:52)
          -> /workspace/node_modules/@grpc/grpc-js/src/client.ts:338:36
      at /workspace/node_modules/@grpc/grpc-js/build/src/call-stream.js:80:35
          -> /workspace/node_modules/@grpc/grpc-js/src/call-stream.ts:170:27
      at Object.onReceiveStatus (node_modules/grpc-gcp/build/src/index.js:73:29)
          -> /workspace/node_modules/grpc-gcp/src/index.ts:129:15
      at InterceptingListenerImpl.onReceiveStatus (node_modules/@grpc/grpc-js/build/src/call-stream.js:75:23)
          -> /workspace/node_modules/@grpc/grpc-js/src/call-stream.ts:166:19
      at Object.onReceiveStatus (node_modules/@grpc/grpc-js/build/src/client-interceptors.js:336:141)
          -> /workspace/node_modules/@grpc/grpc-js/src/client-interceptors.ts:426:34
      at Object.onReceiveStatus (node_modules/@grpc/grpc-js/build/src/client-interceptors.js:299:181)
          -> /workspace/node_modules/@grpc/grpc-js/src/client-interceptors.ts:389:48
      at /workspace/node_modules/@grpc/grpc-js/build/src/call-stream.js:145:78
          -> /workspace/node_modules/@grpc/grpc-js/src/call-stream.ts:276:24
      at processTicksAndRejections (internal/process/task_queues.js:79:11)
```

olavloite committed:
test: add test for deeply nested objects

skuruppu commented on test/codec.ts:
```diff
@@ -631,6 +659,48 @@ describe('codec', () => {
+      );
+    });
+
+    it('should encode deeply-nested object as JSON', () => {
```
Should we also add a case for decoding a deeply-nested object or is that overkill?

olavloite replied:
I think it's reasonable to have it symmetrical, so I've added a test for that as well.

olavloite committed:
test: add test for decoding complex JSON string

skuruppu commented on test/codec.ts:
```diff
@@ -701,6 +701,27 @@ describe('codec', () => {
+
+    it('should decode deeply-nested object as JSON', () => {
+      // Cloud Spanner accepts a nesting level in a JSON string of at most 100.
+      // This test ensures that the encoder is able to decode such a string.
```
```suggestion
      // This test ensures that the decoder is able to decode such a string.
```

olavloite replied:
Done

olavloite committed:
fix: typo

zoercai commented:
Hi @olavloite, we're getting ready to merge the JSON PRs now for Node, Java, and Python. You can merge this PR as soon as all the tests pass. Could you please rebase and rerun CI? Thanks!

olavloite committed:
fix: run linter

olavloite committed:
fix: update expected query output

olavloite committed:
chore: run linter

olavloite commented:
> Hi @olavloite, we're getting ready to merge the JSON PRs now for Node, Java, and Python. You can merge this PR as soon as all the tests pass. Could you please rebase and rerun CI? Thanks!

The JSON samples and other JSON tests pass, but there is some other error occurring for the system tests at the moment that blocks merging this. That error is not specific to this PR, but happens for multiple (all?) PRs at the moment.

skuruppu commented:
I'm reaching out to the Node.js team to figure out why this suddently started happening.

olavloite committed:
chore: run linter

skuruppu committed:
Merge branch 'master' into json-data-type

skuruppu commented:
@olavloite if you can please take a look at the test failure, that would be great. I ran it a couple of times and got the same error so I gathered it must be due to the new tests we added in this PR.

`ERROR: Error: 6 ALREADY_EXISTS: Row [81] in table Venues already exists`.

olavloite committed:
fix: remove duplicate tag test execution

olavloite committed:
fix: remove duplicate transaction tag test run

olavloite commented:
> @olavloite if you can please take a look at the test failure, that would be great. I ran it a couple of times and got the same error so I gathered it must be due to the new tests we added in this PR.
> 
> `ERROR: Error: 6 ALREADY_EXISTS: Row [81] in table Venues already exists`.

It was caused by a merge/rebase error. The previously added test runs for the tagging feature were also added once more by this PR, which meant that the tagging sample tests were executed twice. I've removed the duplicate test runs, and hopefully that should fix it.

skuruppu commented:
Now asked for help from the Node.js team on how to unblock the OwlBot check.

skuruppu committed:
Merge branch 'master' into json-data-type
